### PR TITLE
feat: user vaults v6

### PR DESCRIPTION
**Pull request by uzielflorida**

### CODE REVIEW

File: src/contracts/UserVaultV6.ts [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        return acc;
+      },
+      [] as typeof nfts,
+    );
+
+    for (let i = 0; i < groupedNfts.length; i++) {
+      const { collection, tokenIds, isOldErc721 } = groupedNfts[i];
+      const deposit = isOldErc721
+        ? await this.depositOldERC721s({ vaultId, collection, tokenIds })
+        : await this.depositERC721s({ vaultId, collection, tokenIds });
```
Comment (franco-gondi) [inline src/contracts/UserVaultV6.ts:96]:
```suggestion
      const deposit = await (isOldErc721? this.depositOldERC721s : this.depositERC721s)({ vaultId, collection, tokenIds });
```

Comment (franco-gondi) [reply src/contracts/UserVaultV6.ts:96]: for more clarity maybe move it to a var

File: src/gondi.ts [hunk 2]
```diff
@@ -950,6 +950,33 @@ export class Gondi {
...
+        tokenId: bigint;
+      }
+    | {
+        isOldErc721?: never;
+        tokenId?: never;
+      }
+  )) {
+    if (isOldErc721) {
+      const erc721 = this.contracts.OldERC721(nftAddress);
+      return areSameAddress(await erc721.read.approvedFor([tokenId]), to);
```
Comment (mishuagopian) [inline src/gondi.ts:975]:
```suggestion
      const tokenApprovedFor = await erc721.read.approvedFor([tokenId]);
      return areSameAddress(tokenApprovedFor, to);
```

File: src/gondi.ts [hunk 3]
```diff
@@ -1022,7 +1089,7 @@ export class Gondi {
 
   async createUserVault({
     nfts,
-    userVaultAddress = this.contracts.UserVaultV5.address,
+    userVaultAddress = this.contracts.UserVaultV6.address,
```
Comment (mishuagopian) [inline src/gondi.ts:1092]: Undo change and add TODO

File: src/gondi.ts [hunk 4]
```diff
@@ -1031,16 +1098,25 @@ export class Gondi {
   }
 
   async depositUserVaultERC721s({
-    userVaultAddress = this.contracts.UserVaultV5.address,
+    userVaultAddress = this.contracts.UserVaultV6.address,
```
Comment (mishuagopian) [inline src/gondi.ts:1101]: Same

File: src/gondi.ts [hunk 5]
```diff
@@ -1031,16 +1098,25 @@ export class Gondi {
...
+    ...data
+  }: {
+    userVaultAddress?: Address;
+  } & Parameters<Contracts['UserVaultV6']['depositERC721s']>[0]) {
+    return this.contracts.UserVault(userVaultAddress).depositOldERC721s(data);
+  }
+
   async burnUserVaultAndWithdraw({
-    userVaultAddress = this.contracts.UserVaultV5.address,
+    userVaultAddress = this.contracts.UserVaultV6.address,
```
Comment (mishuagopian) [inline src/gondi.ts:1119]: Same

File: src/contracts/UserVaultV6.ts [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        const index = acc.findIndex((element) => element.collection === collection);
+        if (index === -1) {
+          acc.push({ collection, tokenIds, isOldErc721 });
+        } else {
+          acc[index].tokenIds = [...acc[index].tokenIds, ...tokenIds];
+        }
+        return acc;
+      },
+      [] as typeof nfts,
+    );
```
Comment (franco-gondi) [inline src/contracts/UserVaultV6.ts:90]: can't this be a map collection -> tokenIds?
```suggestion
    const groupedNfts = nfts.reduce(
      (acc, {collection, tokenIds, isOldErc721}) => {
        if (acc?[collection.id]) {
          acc[collection.id].tokenIds = [...acc[index].tokenIds, ...tokenIds];
        } else {
          acc[collection.id] = { collection, tokenIds, isOldErc721 };
        }
        return acc;
      },
      {} as typeof nfts,
    );
```

File: src/contracts/UserVaultV6.ts [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+    for (let i = 0; i < groupedNfts.length; i++) {
```
Comment (franco-gondi) [inline src/contracts/UserVaultV6.ts:92]:
```suggestion
    for (const [_, {collection, tokenIds, isOldErc721}] of Object.entries(groupedNfts) ) {
```

File: src/contracts/UserVaultV6.ts (hunk 6 above)
Comment (mishuagopian) [reply src/contracts/UserVaultV6.ts]: If we do this we gotta change the iteration below, but I got nothing against it

File: src/contracts/UserVaultV5.ts [hunk 8]
```diff
@@ -51,13 +54,14 @@ export class UserVault extends BaseContract<typeof userVaultABIV5> {
...
           events: events.map((event) => event.args),
+          oldEvents: [],
           ...receipt,
         };
       },
     };
   }
 
-  async createVault(nfts: { collection: Address; tokenIds: bigint[] }[]) {
+  async createVault(nfts: { collection: Address; tokenIds: bigint[]; isOldErc721?: boolean }[]) {
```
Comment (mishuagopian) [inline src/contracts/UserVaultV5.ts:64]: Maybe we should get the type from user vault v6?

Comment (mishuagopian) [reply src/contracts/UserVaultV5.ts:64]: Same for all methods? Just thinking outloud

File: examples/userVaultsWithOldErc721.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+  console.log('Creating vault with nfts...');
+  const { vaultId, receipts } = await gondi.createUserVault({
+    userVaultAddress: USER_VAULT_CONTRACT_V6,
+    nfts: [
+      { collection: ANOTHER_COLLECTION, tokenIds: [testTokenId] },
+      { collection: testCollection.contractAddress, tokenIds: [testTokenId] },
+      { collection: OLD_ERC721_COLLECTION, tokenIds: [OLD_ERC721_TOKEN_ID], isOldErc721: true },
+    ],
+  });
```
Comment (franco-gondi) [inline examples/userVaultsWithOldErc721.ts:34]:
```suggestion
  const gondi = users[1];
  const vaultERC721 = [
      {collection: testCollection, tokenId: testTokenId},
      {collection: ANOTHER_COLLECTION, tokenId: testTokenId}
  ];
  const vaultOldERC721 = [
      {collection: ANOTHER_COLLECTION, tokenId: testTokenId}
      {collection: OLD_ERC721_COLLECTION, tokenId: OLD_ERC721_TOKEN_ID}
  ]
  cons vaultNfts = [
      ...vaultERC721,
      ...vaultOldERC721
  ]
  for ([name, address] of [
  ["TEST_COLLECTION_2", ANOTHER_COLLECTION],
  ["USER_VAULT_CONTRACT_V6", USER_VAULT_CONTRACT_V6],
  ["OLD_ERC721_COLLECTION", OLD_ERC721_COLLECTION],
  ])
  if (!isAddress(ANOTHER_COLLECTION)) {
    throw new Error(`
      ${name} were not provided.
      Please, provide addresses to run user vaults example.
    `);
  }
  for(const {collection} of vaultERC721) {
      await approveNFT(gondi, USER_VAULT_CONTRACT_V6, collection.contractAddress);
  }

  console.log('Creating vault with nfts...');
  const { vaultId, receipts } = await gondi.createUserVault({
    userVaultAddress: USER_VAULT_CONTRACT_V6,
    nfts: vaultNfts,
  });
```

File: src/contracts/UserVaultV6.ts [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+      const { collection, tokenIds, isOldErc721 } = groupedNfts[i];
```
Comment (franco-gondi) [inline src/contracts/UserVaultV6.ts]:
```suggestion
    for (const {collection, tokenIds, isOldErc721} of Object.values(groupedNfts) ) {
```

File: examples/userVaultsWithOldErc721.ts (hunk 9 above)
Comment (uzielflorida) [reply examples/userVaultsWithOldErc721.ts:34]: deleted and added contract var on prev example

**Review by mishuagopian — APPROVED** [review]

**Pull request merged**